File: src/csharp/Intel/Iced/Intel/BlockEncoderInternal/SimpleBranchInstr.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if ENCODER && BLOCK_ENCODER
using System;
using System.Diagnostics;

namespace Iced.Intel.BlockEncoderInternal {
	/// <summary>
	/// Simple branch instruction that only has one code value, eg. loopcc, jrcxz
	/// </summary>
	sealed class SimpleBranchInstr : Instr {
		readonly byte bitness;
		Instruction instruction;
		TargetInstr targetInstr;
		BlockData? pointerData;
		InstrKind instrKind;
		readonly byte shortInstructionSize;
		readonly byte nearInstructionSize;
		readonly byte longInstructionSize;
		readonly byte nativeInstructionSize;
		readonly Code nativeCode;

		enum InstrKind : byte {
			Unchanged,
			Short,
			Near,
			Long,
			Uninitialized,
		}

		public SimpleBranchInstr(BlockEncoder blockEncoder, Block block, in Instruction instruction)
			: base(block, instruction.IP) {
			bitness = (byte)blockEncoder.Bitness;
			this.instruction = instruction;
			instrKind = InstrKind.Uninitialized;

			Instruction instrCopy;

			if (!blockEncoder.FixBranches) {
				instrKind = InstrKind.Unchanged;
				instrCopy = instruction;
				instrCopy.NearBranch64 = 0;
				Size = blockEncoder.GetInstructionSize(instrCopy, 0);
			}
			else {
				instrCopy = instruction;
				instrCopy.NearBranch64 = 0;
				shortInstructionSize = (byte)blockEncoder.GetInstructionSize(instrCopy, 0);

				nativeCode = ToNativeBranchCode(instruction.Code, blockEncoder.Bitness);
				if (nativeCode == instruction.Code)
					nativeInstructionSize = shortInstructionSize;
				else {
					instrCopy = instruction;
					instrCopy.InternalSetCodeNoCheck(nativeCode);
					instrCopy.NearBranch64 = 0;
					nativeInstructionSize = (byte)blockEncoder.GetInstructionSize(instrCopy, 0);
				}

				nearInstructionSize = (byte)(blockEncoder.Bitness switch {
					16 => nativeInstructionSize + 2 + 3,
					32 or 64 => nativeInstructionSize + 2 + 5,
					_ => throw new InvalidOperationException(),
				});

				if (blockEncoder.Bitness == 64) {
					longInstructionSize = (byte)(nativeInstructionSize + 2 + CallOrJmpPointerDataInstructionSize64);
					Size = Math.Max(Math.Max(shortInstructionSize, nearInstructionSize), longInstructionSize);
				}
				else
					Size = Math.Max(shortInstructionSize, nearInstructionSize);
			}
		}

		static Code ToNativeBranchCode(Code code, int bitness) {
			Code c16, c32, c64;
			switch (code) {
			case Code.Loopne_rel8_16_CX:
			case Code.Loopne_rel8_32_CX:
				c16 = Code.Loopne_rel8_16_CX;
				c32 = Code.Loopne_rel8_32_CX;
				c64 = Code.INVALID;
				break;

			case Code.Loopne_rel8_16_ECX:
			case Code.Loopne_rel8_32_ECX:
			case Code.Loopne_rel8_64_ECX:
				c16 = Code.Loopne_rel8_16_ECX;
				c32 = Code.Loopne_rel8_32_ECX;
				c64 = Code.Loopne_rel8_64_ECX;
				break;

			case Code.Loopne_rel8_16_RCX:
			case Code.Loopne_rel8_64_RCX:
				c16 = Code.Loopne_rel8_16_RCX;
				c32 = Code.INVALID;
				c64 = Code.Loopne_rel8_64_RCX;
				break;

			case Code.Loope_rel8_16_CX:
			case Code.Loope_rel8_32_CX:
				c16 = Code.Loope_rel8_16_CX;
				c32 = Code.Loope_rel8_32_CX;
				c64 = Code.INVALID;
				break;

			case Code.Loope_rel8_16_ECX:
			case Code.Loope_rel8_32_ECX:
			case Code.Loope_rel8_64_ECX:
				c16 = Code.Loope_rel8_16_ECX;
				c32 = Code.Loope_rel8_32_ECX;
				c64 = Code.Loope_rel8_64_ECX;
				break;

			case Code.Loope_rel8_16_RCX:
			case Code.Loope_rel8_64_RCX:
				c16 = Code.Loope_rel8_16_RCX;
				c32 = Code.INVALID;
				c64 = Code.Loope_rel8_64_RCX;
				break;

			case Code.Loop_rel8_16_CX:
			case Code.Loop_rel8_32_CX:
				c16 = Code.Loop_rel8_16_CX;
				c32 = Code.Loop_rel8_32_CX;
				c64 = Code.INVALID;
				break;

			case Code.Loop_rel8_16_ECX:
			case Code.Loop_rel8_32_ECX:
			case Code.Loop_rel8_64_ECX:
				c16 = Code.Loop_rel8_16_ECX;
				c32 = Code.Loop_rel8_32_ECX;
				c64 = Code.Loop_rel8_64_ECX;
				break;

			case Code.Loop_rel8_16_RCX:
			case Code.Loop_rel8_64_RCX:
				c16 = Code.Loop_rel8_16_RCX;
				c32 = Code.INVALID;
				c64 = Code.Loop_rel8_64_RCX;
				break;

			case Code.Jcxz_rel8_16:
			case Code.Jcxz_rel8_32:
				c16 = Code.Jcxz_rel8_16;
				c32 = Code.Jcxz_rel8_32;
				c64 = Code.INVALID;
				break;

			case Code.Jecxz_rel8_16:
			case Code.Jecxz_rel8_32:
			case Code.Jecxz_rel8_64:
				c16 = Code.Jecxz_rel8_16;
				c32 = Code.Jecxz_rel8_32;
				c64 = Code.Jecxz_rel8_64;
				break;

			case Code.Jrcxz_rel8_16:
			case Code.Jrcxz_rel8_64:
				c16 = Code.Jrcxz_rel8_16;
				c32 = Code.INVALID;
				c64 = Code.Jrcxz_rel8_64;
				break;

			default:
				throw new ArgumentOutOfRangeException(nameof(code));
			}

			return bitness switch {
				16 => c16,
				32 => c32,
				64 => c64,
				_ => throw new ArgumentOutOfRangeException(nameof(bitness)),
			};
		}

		public override void Initialize(BlockEncoder blockEncoder) =>
			targetInstr = blockEncoder.GetTarget(instruction.NearBranchTarget);

		public override bool Optimize(ulong gained) => TryOptimize(gained);

		bool TryOptimize(ulong gained) {
			if (instrKind == InstrKind.Unchanged || instrKind == InstrKind.Short) {
				Done = true;
				return false;
			}

			var targetAddress = targetInstr.GetAddress();
			var nextRip = IP + shortInstructionSize;
			long diff = (long)(targetAddress - nextRip);
			diff = ConvertDiffToBitnessDiff(bitness, CorrectDiff(targetInstr.IsInBlock(Block), diff, gained));
			if (sbyte.MinValue <= diff && diff <= sbyte.MaxValue) {
				if (pointerData is not null)
					pointerData.IsValid = false;
				instrKind = InstrKind.Short;
				Size = shortInstructionSize;
				Done = true;
				return true;
			}

			// If it's in the same block, we assume the target is at most 2GB away.
			bool useNear = bitness != 64 || targetInstr.IsInBlock(Block);
			if (!useNear) {
				targetAddress = targetInstr.GetAddress();
				nextRip = IP + nearInstructionSize;
				diff = (long)(targetAddress - nextRip);
				diff = ConvertDiffToBitnessDiff(bitness, CorrectDiff(targetInstr.IsInBlock(Block), diff, gained));
				useNear = int.MinValue <= diff && diff <= int.MaxValue;
			}
			if (useNear) {
				if (pointerData is not null)
					pointerData.IsValid = false;
				if (diff < (long)IcedConstants.MaxInstructionLength * sbyte.MinValue ||
					diff > (long)IcedConstants.MaxInstructionLength * sbyte.MaxValue) {
					Done = true;
				}
				instrKind = InstrKind.Near;
				Size = nearInstructionSize;
				return true;
			}

			if (pointerData is null)
				pointerData = Block.AllocPointerLocation();
			instrKind = InstrKind.Long;
			return false;
		}

		public override string? TryEncode(Encoder encoder, out ConstantOffsets constantOffsets, out bool isOriginalInstruction) {
			string? errorMessage;
			Instruction instr;
			uint size;
			uint instrLen;
			switch (instrKind) {
			case InstrKind.Unchanged:
			case InstrKind.Short:
				isOriginalInstruction = true;
				instruction.NearBranch64 = targetInstr.GetAddress();
				if (!encoder.TryEncode(instruction, IP, out _, out errorMessage)) {
					constantOffsets = default;
					return CreateErrorMessage(errorMessage, instruction);
				}
				constantOffsets = encoder.GetConstantOffsets();
				return null;

			case InstrKind.Near:
				isOriginalInstruction = false;
				constantOffsets = default;

				// Code:
				//		brins tmp		; nativeInstructionSize
				//		jmp short skip	; 2
				//	tmp:
				//		jmp near target	; 3/5/5
				//	skip:

				instr = instruction;
				instr.InternalSetCodeNoCheck(nativeCode);
				instr.NearBranch64 = IP + nativeInstructionSize + 2;
				if (!encoder.TryEncode(instr, IP, out size, out errorMessage))
					return CreateErrorMessage(errorMessage, instruction);

				instr = new Instruction();
				instr.NearBranch64 = IP + nearInstructionSize;
				Code codeNear;
				switch (encoder.Bitness) {
				case 16:
					instr.InternalSetCodeNoCheck(Code.Jmp_rel8_16);
					codeNear = Code.Jmp_rel16;
					instr.Op0Kind = OpKind.NearBranch16;
					break;

				case 32:
					instr.InternalSetCodeNoCheck(Code.Jmp_rel8_32);
					codeNear = Code.Jmp_rel32_32;
					instr.Op0Kind = OpKind.NearBranch32;
					break;

				case 64:
					instr.InternalSetCodeNoCheck(Code.Jmp_rel8_64);
					codeNear = Code.Jmp_rel32_64;
					instr.Op0Kind = OpKind.NearBranch64;
					break;

				default:
					throw new InvalidOperationException();
				}
				if (!encoder.TryEncode(instr, IP + size, out instrLen, out errorMessage))
					return CreateErrorMessage(errorMessage, instruction);
				size += instrLen;

				instr.InternalSetCodeNoCheck(codeNear);
				instr.NearBranch64 = targetInstr.GetAddress();
				encoder.TryEncode(instr, IP + size, out instrLen, out errorMessage);
				if (errorMessage is not null)
					return CreateErrorMessage(errorMessage, instruction);
				return null;

			case InstrKind.Long:
				Debug.Assert(encoder.Bitness == 64);
				Debug2.Assert(pointerData is not null);
				isOriginalInstruction = false;
				constantOffsets = default;
				pointerData.Data = targetInstr.GetAddress();

				// Code:
				//		brins tmp		; nativeInstructionSize
				//		jmp short skip	; 2
				//	tmp:
				//		jmp [mem_loc]	; 6
				//	skip:

				instr = instruction;
				instr.InternalSetCodeNoCheck(nativeCode);
				instr.NearBranch64 = IP + nativeInstructionSize + 2;
				if (!encoder.TryEncode(instr, IP, out instrLen, out errorMessage))
					return CreateErrorMessage(errorMessage, instruction);
				size = instrLen;

				instr = new Instruction();
				instr.NearBranch64 = IP + longInstructionSize;
				switch (encoder.Bitness) {
				case 16:
					instr.InternalSetCodeNoCheck(Code.Jmp_rel8_16);
					instr.Op0Kind = OpKind.NearBranch16;
					break;

				case 32:
					instr.InternalSetCodeNoCheck(Code.Jmp_rel8_32);
					instr.Op0Kind = OpKind.NearBranch32;
					break;

				case 64:
					instr.InternalSetCodeNoCheck(Code.Jmp_rel8_64);
					instr.Op0Kind = OpKind.NearBranch64;
					break;

				default:
					throw new InvalidOperationException();
				}
				if (!encoder.TryEncode(instr, IP + size, out instrLen, out errorMessage))
					return CreateErrorMessage(errorMessage, instruction);
				size += instrLen;

				errorMessage = EncodeBranchToPointerData(encoder, isCall: false, IP + size, pointerData, out _, Size - size);
				if (errorMessage is not null)
					return CreateErrorMessage(errorMessage, instruction);
				return null;

			case InstrKind.Uninitialized:
			default:
				throw new InvalidOperationException();
			}
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/enc/SimpleBranchInstr.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.enc;

import com.github.icedland.iced.x86.Code;
import com.github.icedland.iced.x86.ConstantOffsets;
import com.github.icedland.iced.x86.Instruction;
import com.github.icedland.iced.x86.OpKind;
import com.github.icedland.iced.x86.internal.IcedConstants;

final class SimpleBranchInstr extends Instr {
	private final byte bitness;
	private Instruction instruction;
	private TargetInstr targetInstr;
	private BlockData pointerData;
	private byte instrKind;
	private final byte shortInstructionSize;
	private final byte nearInstructionSize;
	private final byte longInstructionSize;
	private final byte nativeInstructionSize;
	private final short nativeCode;

	private static final class InstrKind {
		static final byte UNCHANGED = 0;
		static final byte SHORT = 1;
		static final byte NEAR = 2;
		static final byte LONG = 3;
		static final byte UNINITIALIZED = 4;
	}

	public SimpleBranchInstr(BlockEncoder blockEncoder, Block block, Instruction instruction) {
		super(block, instruction.getIP());
		this.bitness = (byte)blockEncoder.getBitness();
		this.instruction = instruction;
		instrKind = InstrKind.UNINITIALIZED;

		Instruction instrCopy;

		if (!blockEncoder.fixBranches()) {
			instrKind = InstrKind.UNCHANGED;
			instrCopy = instruction.copy();
			instrCopy.setNearBranch64(0);
			size = blockEncoder.getInstructionSize(instrCopy, 0);
			shortInstructionSize = 0;
			nearInstructionSize = 0;
			longInstructionSize = 0;
			nativeInstructionSize = 0;
			nativeCode = 0;
		}
		else {
			instrCopy = instruction.copy();
			instrCopy.setNearBranch64(0);
			shortInstructionSize = (byte)blockEncoder.getInstructionSize(instrCopy, 0);

			nativeCode = (short)toNativeBranchCode(instruction.getCode(), blockEncoder.getBitness());
			if (nativeCode == instruction.getCode())
				nativeInstructionSize = shortInstructionSize;
			else {
				instrCopy = instruction.copy();
				instrCopy.setCode(nativeCode & 0xFFFF);
				instrCopy.setNearBranch64(0);
				nativeInstructionSize = (byte)blockEncoder.getInstructionSize(instrCopy, 0);
			}

			switch (blockEncoder.getBitness()) {
			case 16:
				nearInstructionSize = (byte)(nativeInstructionSize + 2 + 3);
				break;
			case 32:
			case 64:
				nearInstructionSize = (byte)(nativeInstructionSize + 2 + 5);
				break;
			default:
				throw new UnsupportedOperationException();
			}

			if (blockEncoder.getBitness() == 64) {
				longInstructionSize = (byte)(nativeInstructionSize + 2 + CALL_OR_JMP_POINTER_DATA_INSTRUCTION_SIZE64);
				size = Math.max(Math.max(shortInstructionSize, nearInstructionSize), longInstructionSize);
			}
			else {
				longInstructionSize = 0;
				size = Math.max(shortInstructionSize, nearInstructionSize);
			}
		}
	}

	static int toNativeBranchCode(int code, int bitness) {
		int c16, c32, c64;
		switch (code) {
		case Code.LOOPNE_REL8_16_CX:
		case Code.LOOPNE_REL8_32_CX:
			c16 = Code.LOOPNE_REL8_16_CX;
			c32 = Code.LOOPNE_REL8_32_CX;
			c64 = Code.INVALID;
			break;

		case Code.LOOPNE_REL8_16_ECX:
		case Code.LOOPNE_REL8_32_ECX:
		case Code.LOOPNE_REL8_64_ECX:
			c16 = Code.LOOPNE_REL8_16_ECX;
			c32 = Code.LOOPNE_REL8_32_ECX;
			c64 = Code.LOOPNE_REL8_64_ECX;
			break;

		case Code.LOOPNE_REL8_16_RCX:
		case Code.LOOPNE_REL8_64_RCX:
			c16 = Code.LOOPNE_REL8_16_RCX;
			c32 = Code.INVALID;
			c64 = Code.LOOPNE_REL8_64_RCX;
			break;

		case Code.LOOPE_REL8_16_CX:
		case Code.LOOPE_REL8_32_CX:
			c16 = Code.LOOPE_REL8_16_CX;
			c32 = Code.LOOPE_REL8_32_CX;
			c64 = Code.INVALID;
			break;

		case Code.LOOPE_REL8_16_ECX:
		case Code.LOOPE_REL8_32_ECX:
		case Code.LOOPE_REL8_64_ECX:
			c16 = Code.LOOPE_REL8_16_ECX;
			c32 = Code.LOOPE_REL8_32_ECX;
			c64 = Code.LOOPE_REL8_64_ECX;
			break;

		case Code.LOOPE_REL8_16_RCX:
		case Code.LOOPE_REL8_64_RCX:
			c16 = Code.LOOPE_REL8_16_RCX;
			c32 = Code.INVALID;
			c64 = Code.LOOPE_REL8_64_RCX;
			break;

		case Code.LOOP_REL8_16_CX:
		case Code.LOOP_REL8_32_CX:
			c16 = Code.LOOP_REL8_16_CX;
			c32 = Code.LOOP_REL8_32_CX;
			c64 = Code.INVALID;
			break;

		case Code.LOOP_REL8_16_ECX:
		case Code.LOOP_REL8_32_ECX:
		case Code.LOOP_REL8_64_ECX:
			c16 = Code.LOOP_REL8_16_ECX;
			c32 = Code.LOOP_REL8_32_ECX;
			c64 = Code.LOOP_REL8_64_ECX;
			break;

		case Code.LOOP_REL8_16_RCX:
		case Code.LOOP_REL8_64_RCX:
			c16 = Code.LOOP_REL8_16_RCX;
			c32 = Code.INVALID;
			c64 = Code.LOOP_REL8_64_RCX;
			break;

		case Code.JCXZ_REL8_16:
		case Code.JCXZ_REL8_32:
			c16 = Code.JCXZ_REL8_16;
			c32 = Code.JCXZ_REL8_32;
			c64 = Code.INVALID;
			break;

		case Code.JECXZ_REL8_16:
		case Code.JECXZ_REL8_32:
		case Code.JECXZ_REL8_64:
			c16 = Code.JECXZ_REL8_16;
			c32 = Code.JECXZ_REL8_32;
			c64 = Code.JECXZ_REL8_64;
			break;

		case Code.JRCXZ_REL8_16:
		case Code.JRCXZ_REL8_64:
			c16 = Code.JRCXZ_REL8_16;
			c32 = Code.INVALID;
			c64 = Code.JRCXZ_REL8_64;
			break;

		default:
			throw new IllegalArgumentException("code");
		}

		switch (bitness) {
		case 16:
			return c16;
		case 32:
			return c32;
		case 64:
			return c64;
		default:
			throw new IllegalArgumentException("bitness");
		}
	}

	@Override
	void initialize(BlockEncoder blockEncoder) {
		targetInstr = blockEncoder.getTarget(instruction.getNearBranchTarget());
	}

	@Override
	boolean optimize(long gained) {
		return tryOptimize(gained);
	}

	private boolean tryOptimize(long gained) {
		if (instrKind == InstrKind.UNCHANGED || instrKind == InstrKind.SHORT) {
			done = true;
			return false;
		}

		long targetAddress = targetInstr.getAddress();
		long nextRip = ip + shortInstructionSize;
		long diff = targetAddress - nextRip;
		diff = convertDiffToBitnessDiff(bitness, correctDiff(targetInstr.isInBlock(block), diff, gained));
		if (-0x80 <= diff && diff <= 0x7F) {
			if (pointerData != null)
				pointerData.isValid = false;
			instrKind = InstrKind.SHORT;
			size = shortInstructionSize;
			done = true;
			return true;
		}

		// If it's in the same block, we assume the target is at most 2GB away.
		boolean useNear = bitness != 64 || targetInstr.isInBlock(block);
		if (!useNear) {
			targetAddress = targetInstr.getAddress();
			nextRip = ip + nearInstructionSize;
			diff = targetAddress - nextRip;
			diff = convertDiffToBitnessDiff(bitness, correctDiff(targetInstr.isInBlock(block), diff, gained));
			useNear = -0x8000_0000 <= diff && diff <= 0x7FFF_FFFF;
		}
		if (useNear) {
			if (pointerData != null)
				pointerData.isValid = false;
			if (diff < IcedConstants.MAX_INSTRUCTION_LENGTH * -0x80 || diff > IcedConstants.MAX_INSTRUCTION_LENGTH * 0x7F) {
				done = true;
			}
			instrKind = InstrKind.NEAR;
			size = nearInstructionSize;
			return true;
		}

		if (pointerData == null)
			pointerData = block.allocPointerLocation();
		instrKind = InstrKind.LONG;
		return false;
	}

	@Override
	String tryEncode(Encoder encoder, TryEncodeResult result) {
		Object encResult;
		Instruction instr;
		int size;
		switch (instrKind) {
		case InstrKind.UNCHANGED:
		case InstrKind.SHORT:
			result.isOriginalInstruction = true;
			instr = instruction.copy();
			instr.setNearBranch64(targetInstr.getAddress());
			encResult = encoder.tryEncode(instr, ip);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instr);
			result.constantOffsets = encoder.getConstantOffsets();
			return null;

		case InstrKind.NEAR:
			result.isOriginalInstruction = false;
			result.constantOffsets = new ConstantOffsets();

			// Code:
			//		brins tmp		; nativeInstructionSize
			//		jmp short skip	; 2
			//	tmp:
			//		jmp near target	; 3/5/5
			//	skip:

			instr = instruction.copy();
			instr.setCode(nativeCode & 0xFFFF);
			instr.setNearBranch64(ip + nativeInstructionSize + 2);
			encResult = encoder.tryEncode(instr, ip);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instruction);
			size = ((Integer)encResult).intValue();

			instr = new Instruction();
			instr.setNearBranch64(ip + nearInstructionSize);
			int codeNear;
			switch (encoder.getBitness()) {
			case 16:
				instr.setCode(Code.JMP_REL8_16);
				codeNear = Code.JMP_REL16;
				instr.setOp0Kind(OpKind.NEAR_BRANCH16);
				break;

			case 32:
				instr.setCode(Code.JMP_REL8_32);
				codeNear = Code.JMP_REL32_32;
				instr.setOp0Kind(OpKind.NEAR_BRANCH32);
				break;

			case 64:
				instr.setCode(Code.JMP_REL8_64);
				codeNear = Code.JMP_REL32_64;
				instr.setOp0Kind(OpKind.NEAR_BRANCH64);
				break;

			default:
				throw new UnsupportedOperationException();
			}
			encResult = encoder.tryEncode(instr, ip + size);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instruction);
			size += ((Integer)encResult).intValue();

			instr.setCode(codeNear);
			instr.setNearBranch64(targetInstr.getAddress());
			encResult = encoder.tryEncode(instr, ip + size);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instruction);
			return null;

		case InstrKind.LONG:
			assert encoder.getBitness() == 64 : encoder.getBitness();
			assert pointerData != null;
			result.isOriginalInstruction = false;
			result.constantOffsets = new ConstantOffsets();
			pointerData.data = targetInstr.getAddress();

			// Code:
			//		brins tmp		; nativeInstructionSize
			//		jmp short skip	; 2
			//	tmp:
			//		jmp [mem_loc]	; 6
			//	skip:

			instr = instruction.copy();
			instr.setCode(nativeCode & 0xFFFF);
			instr.setNearBranch64(ip + nativeInstructionSize + 2);
			encResult = encoder.tryEncode(instr, ip);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instruction);
			size = ((Integer)encResult).intValue();

			instr = new Instruction();
			instr.setNearBranch64(ip + longInstructionSize);
			switch (encoder.getBitness()) {
			case 16:
				instr.setCode(Code.JMP_REL8_16);
				instr.setOp0Kind(OpKind.NEAR_BRANCH16);
				break;

			case 32:
				instr.setCode(Code.JMP_REL8_32);
				instr.setOp0Kind(OpKind.NEAR_BRANCH32);
				break;

			case 64:
				instr.setCode(Code.JMP_REL8_64);
				instr.setOp0Kind(OpKind.NEAR_BRANCH64);
				break;

			default:
				throw new UnsupportedOperationException();
			}
			encResult = encoder.tryEncode(instr, ip + size);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instruction);
			size += ((Integer)encResult).intValue();

			encResult = encodeBranchToPointerData(encoder, false, ip + size, pointerData, this.size - size);
			if (encResult instanceof String)
				return createErrorMessage((String)encResult, instruction);
			return null;

		case InstrKind.UNINITIALIZED:
		default:
			throw new UnsupportedOperationException();
		}
	}
}
